carma_cloud_client/src/carma_cloud_client_node.cpp

```diff
@@ -0,0 +1,199 @@
+
+    result.successful = !error;
+
+    return result;
+  }
+
+  carma_ros2_utils::CallbackReturn CarmaCloudClient::handle_on_configure(const rclcpp_lifecycle::State &)
+  {
+    RCLCPP_INFO_STREAM(this->get_logger(), "CarmaCloudClient trying to configure");
+    // Reset config
+    config_ = Config();
+
+    // Load parameters
+    get_parameter<std::string>("url", config_.url);
+    get_parameter<std::string>("base_hb", config_.base_hb);
+    get_parameter<std::string>("base_req", config_.base_req);
+    get_parameter<std::string>("base_ack", config_.base_ack);
+    get_parameter<std::string>("port", config_.port);
+    get_parameter<std::string>("list", config_.list);
+    get_parameter<std::string>("method", config_.method);
+    get_parameter<int>("fetchtime", config_.fetchtime);
+
+    // Register runtime parameter update callback
+    add_on_set_parameters_callback(std::bind(&CarmaCloudClient::parameter_update_callback, this, std_ph::_1));
+
+    // Setup subscribers
+    tcr_sub_ = create_subscription<carma_v2x_msgs::msg::TrafficControlRequest>("outgoing_geofence_request", 10,
+                                                              std::bind(&CarmaCloudClient::tcr_callback, this, std_ph::_1));                                                     
+
+
+    // Return success if everything initialized successfully
+    return CallbackReturn::SUCCESS;
+  }
+
+  void CarmaCloudClient::tcr_callback(carma_v2x_msgs::msg::TrafficControlRequest::UniquePtr msg)
+  {
+    carma_v2x_msgs::msg::TrafficControlRequest request_msg(*msg.get());
+
+    char xml_str[10000]; 
+
+    XMLconversion(xml_str, request_msg);
+
+    CloudSend(xml_str, config_.url, config_.base_req, config_.method);
+    
+    RCLCPP_INFO_STREAM(  get_logger(), "tcr_sub_ callback called ");
+  }
+
+  void CarmaCloudClient::XMLconversion(char* xml_str, carma_v2x_msgs::msg::TrafficControlRequest request_msg)
+  {
+    j2735_v2x_msgs::msg::TrafficControlRequest j2735_tcr;
+
+    j2735_convertor::geofence_request::convert(request_msg, j2735_tcr);
+
+    RCLCPP_INFO_STREAM(  get_logger(), "converted: "); 
+
+    size_t hexlen = 2; //size of each hex representation with a leading 0
+    char reqid[j2735_tcr.tcr_v01.reqid.id.size() * hexlen + 1];
+    for (int i = 0; i < j2735_tcr.tcr_v01.reqid.id.size(); i++)
+    {
+      sprintf(reqid+(i*hexlen), "%.2X", j2735_tcr.tcr_v01.reqid.id[i]);
+    }
+
+	  RCLCPP_INFO_STREAM(  get_logger(), "reqid: " << reqid);
+  
+    long int reqseq = j2735_tcr.tcr_v01.reqseq;
+    RCLCPP_INFO_STREAM(  get_logger(), "reqseq: " << reqseq);
+
+	  long int scale = j2735_tcr.tcr_v01.scale;
+    RCLCPP_INFO_STREAM(  get_logger(), "scale: " << scale);
+
+    int totBounds =  j2735_tcr.tcr_v01.bounds.size();
+    int cnt=0;
+    char bounds_str[5000];
+		strcpy(bounds_str,"");
+
+    //  get current time 
+    std::time_t tm = this->now().seconds()/60 - config_.fetchtime*24*60; //  T minus fetchtime*24 hours in  min  
+
+    while(cnt<totBounds)
```

Comment:
not sure why do we need a while loop?

Reply:
I used the logic for converting TCR to XML from the carma cloud plugin code in v2x hub